### DISCUSSION TITLE
Change from split() to vector_split()

### DISCUSSION THREAD

**Discussion by MaurusGubser** [lawmurray/Birch#15] [Q&A]
I use Birch for implementing a HMM for a model of two legs in a course of motion. Basically, I have a state transition model, which is linear and Gaussian with a non-diagonal covariance and an observation model, which is non-linear with Gaussian noise. Thus, I use a vector expression for the state variable, which I split up in a vector of expressions to compute the observation variable.

Previously, I used _split_ in my code. After the most recent update of Birch, split is deprecated. I guess I should use _vector_split_, but it is not clear to me, how a _VectorSplitExpression_ is used. With _split_, I could define a new variable _x_ (vector of expressions), whose elements I could access using the bracket operator, see code fragment below.

```
function state_to_observation(u:Random<Real[_]>) -> Array<Expression<Real>> {
  let x <- split<Real>(u);
  y:Array<Expression<Real>>;

  y.insert(1, x[15]*cst_0 + x[13]*cos(x[3]) + (x[14] + g)*sin(x[3]));
  ...
  return y;
}
```

How can I use _vector_split_ and how can I access the elements of the vector after using _vector_split_?

Not sure if this is the right place to ask a question, but I did not find any other place to do so.

**Comment by lawmurray** [comment]
Thanks for the question. The aim has been to make the use of `split` unnecessary in circumstances such as this, so that you can just access elements directly with a new function `element()`. Can you try something like this?
```
function state_to_observation(u:Random<Real[_]>) -> Array<Expression<Real>> {
  y:Array<Expression<Real>>;
  y.insert(1, element(u, 15)*cst_0 + element(u, 13)*cos(element(u, 3)) + (element(u, 14) + g)*sin(element(u, 3));
  ...
  return y;
}
```
On the to-do list is to overload the square brackets again so that e.g. `element(u, 15)` can be replaced with (the nicer) `u[15]` again.

(Footnote: No worries posting this as an issue, have been meaning to open the discussion section in GitHub, so have just done so now and moved this over there.)

**Reply by MaurusGubser** [reply]
Great, thanks a lot for your fast answer, ~~that solved my issue~~. Sorry, I was not aware of the `element` function. 

By the way, I'm really pleased working with Birch, it's a great tool.

Edit: I was too hasty. There was still an issue. Using your suggestion, I got the following error:

```
error: cannot convert ‘VectorElement<VectorSplitExpression<Real[_], Random<Real[_] > >, Integer>’ to ‘Expression<Real>’
src/container/Array.birch:144: note:   initializing argument 2 of ‘Array<Type>.insert(Integer, Type) [with Type = Expression<Real>; Integer = Integer]’
```
I could work around this using the following:
```
function state_to_observation(u:Random<Real[_]>) -> Array<Expression<Real>> {
  x:Array<Expression<Real>>;
  x.insert(1, box(element(u, 15))*3 + box(element(u, 13))*cos(box(element(u, 3))) + (box(element(u, 14)) + 9.81)*sin(box(element(u, 3))));
  ...
  return x;
}
```
But I'm not sure, if that is the intended way to do this.

**Reply by lawmurray** [reply]
You have the right idea. Here, I think you can just `box` the whole expression actually:
```
x.insert(1, box(element(u, 15)*3 + element(u, 13)*cos(element(u, 3)) + (element(u, 14) + 9.81)*sin(element(u, 3)));
```
What's going on here is that `box` creates a new `Expression` object each time you use it. That can get slow: objects are relatively expensive as they are heap-allocated and reference-counted. But they are also very useful, of course, as in storing them in an `Array` here. We've been updating lazy expressions (including the earlier change with `split`) to reduce the number of objects created as well as reveal many more optimization opportunities to the C++ compiler by static typing, which has produced some big performance gains recently.

**Reply by MaurusGubser** [reply]
Thanks, that worked. For now, the slower performance is not (yet) my main problem; maybe later, I need to re-think about how I implement the model to avoid creating too many objects. I'm sure, concerning performance, there is a lot to improve in my implementation.
If you're interested and it is of any benefit, I can share my model as an example, as soon as it works well.

**Reply by lawmurray** [reply]
Missed your last comment. Indeed, please share out when you're happy with it, new examples are always welcome.
